File: web/src/entities/groundWalkingMonster.ts
import { Direction } from '../constants.js';
import { ImmutableVec2 } from '../math.js';
import { lerp } from '../utils.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import { Sound } from '../media/sound.js';
import type { Animation } from '../media/animation.js';
import { collides } from '../collision.js';

const enum Facing { LEFT_UP, RIGHT_DOWN }

export const enum GroundWalkingMonsterType { FLOOR, LEFT_WALL, RIGHT_WALL, ROOF }

const MIN_WALKING_SPEED = 40.0;
const MAX_WALKING_SPEED = 70.0;

export class GroundWalkingMonster extends Entity {
  private readonly animation: Animation;
  private facing: Facing = Facing.RIGHT_DOWN;
  private frame = 0;
  private readonly type: GroundWalkingMonsterType;

  constructor(type: GroundWalkingMonsterType) {
    super();
    this.type = type;
    this.setSize(new ImmutableVec2(10, 10));
    this.animation = Resource.getAnimation('data/images/saw.bmp', 2);
  }

  getLayer(): number { return 1; }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const x = Math.round(this.getPosition().x - this.animation.getFrameWidth() / 2 + offsetX);
    const y = Math.round(this.getPosition().y - this.animation.getFrameHeight() / 2 + offsetY);
    this.animation.drawFrame(ctx, Math.floor(this.frame / 3), x, y);
  }

  override update(): void {
    const hero = this.myRoom.getHero();
    const speed = lerp(MIN_WALKING_SPEED, MAX_WALKING_SPEED, PlayerSkill.get());

    if (this.type === GroundWalkingMonsterType.FLOOR || this.type === GroundWalkingMonsterType.ROOF) {
      if (this.facing === Facing.LEFT_UP) this.getVelocity().set(-speed, 0);
      else this.getVelocity().set(speed, 0);
    } else {
      if (this.facing === Facing.LEFT_UP) this.getVelocity().set(0, -speed);
      else this.getVelocity().set(0, speed);
    }

    const bumps = this.moveWithCollision();

    if (bumps.has(Direction.LEFT) || bumps.has(Direction.UP)) this.facing = Facing.RIGHT_DOWN;
    if (bumps.has(Direction.RIGHT) || bumps.has(Direction.DOWN)) this.facing = Facing.LEFT_UP;

    let offsetX: number;
    let offsetY: number;

    if (this.type === GroundWalkingMonsterType.FLOOR || this.type === GroundWalkingMonsterType.ROOF) {
      offsetX = this.facing === Facing.LEFT_UP ? -this.getHalfSize().x - 2 : this.getHalfSize().x + 2;
      offsetY = this.type === GroundWalkingMonsterType.FLOOR ? this.getHalfSize().y + 2 : -this.getHalfSize().y - 2;
    } else {
      offsetX = this.type === GroundWalkingMonsterType.LEFT_WALL ? -this.getHalfSize().x - 2 : this.getHalfSize().x + 2;
      offsetY = this.facing === Facing.RIGHT_DOWN ? this.getHalfSize().y + 2 : -this.getHalfSize().y - 2;
    }

    const tx = Math.floor((this.getPosition().x + offsetX) / this.myRoom.getTileWidth());
    const ty = Math.floor((this.getPosition().y + offsetY) / this.myRoom.getTileHeight());

    if (!this.myRoom.isCollidable(tx, ty)) {
      this.facing = this.facing === Facing.LEFT_UP ? Facing.RIGHT_DOWN : Facing.LEFT_UP;
    }

    if (hero.Collides(this)) hero.kill();

    if (hero.hasHook() && collides(this, hero.getHookCollidable())) {
      Sound.playSample('data/sounds/hook');
      hero.detachHook();
    }

    this.frame++;
  }
}

File: wii/src/entities/ground_walking_monster.cpp
#include "entities/ground_walking_monster.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "media/sound.h"
#include "room.h"
#include "player_skill.h"
#include "collision.h"
#include "utils.h"
#include <cmath>

static const float MIN_WALKING_SPEED = 40.0f;
static const float MAX_WALKING_SPEED = 70.0f;

GroundWalkingMonster::GroundWalkingMonster(GroundWalkingMonsterType type)
    : type_(type) {
    setSize(Vec2(10, 10));
    animation_ = Resource::getAnimation("data/images/saw.bmp", 2);
}

void GroundWalkingMonster::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(position_.x)) - animation_->getFrameWidth() / 2 + offsetX;
    int y = static_cast<int>(std::round(position_.y)) - animation_->getFrameHeight() / 2 + offsetY;
    animation_->drawFrame(renderer, frame_ / 3, x, y);
}

void GroundWalkingMonster::update() {
    Hero* hero = room_->getHero();
    float speed = lerp(MIN_WALKING_SPEED, MAX_WALKING_SPEED, PlayerSkill::get());

    if (type_ == GroundWalkingMonsterType::FLOOR || type_ == GroundWalkingMonsterType::ROOF) {
        if (facing_ == Facing::LEFT_UP) velocity_ = Vec2(-speed, 0);
        else velocity_ = Vec2(speed, 0);
    } else {
        if (facing_ == Facing::LEFT_UP) velocity_ = Vec2(0, -speed);
        else velocity_ = Vec2(0, speed);
    }

    auto bumps = moveWithCollision();

    if (bumps.count(Direction::LEFT) || bumps.count(Direction::UP))
        facing_ = Facing::RIGHT_DOWN;
    if (bumps.count(Direction::RIGHT) || bumps.count(Direction::DOWN))
        facing_ = Facing::LEFT_UP;

    float offsetX, offsetY;

    if (type_ == GroundWalkingMonsterType::FLOOR || type_ == GroundWalkingMonsterType::ROOF) {
        offsetX = (facing_ == Facing::LEFT_UP) ? -halfSize_.x - 2 : halfSize_.x + 2;
        offsetY = (type_ == GroundWalkingMonsterType::FLOOR) ? halfSize_.y + 2 : -halfSize_.y - 2;
    } else {
        offsetX = (type_ == GroundWalkingMonsterType::LEFT_WALL) ? -halfSize_.x - 2 : halfSize_.x + 2;
        offsetY = (facing_ == Facing::RIGHT_DOWN) ? halfSize_.y + 2 : -halfSize_.y - 2;
    }

    int tx = static_cast<int>(std::floor((position_.x + offsetX) / room_->getTileWidth()));
    int ty = static_cast<int>(std::floor((position_.y + offsetY) / room_->getTileHeight()));

    if (!room_->isCollidable(tx, ty)) {
        facing_ = (facing_ == Facing::LEFT_UP) ? Facing::RIGHT_DOWN : Facing::LEFT_UP;
    }

    if (hero->Collides(*this)) hero->kill();

    if (hero->hasHook()) {
        SimpleCollidable hookCol = hero->getHookCollidable();
        if (collides(*this, hookCol)) {
            Sound::playSample("data/sounds/hook");
            hero->detachHook();
        }
    }

    frame_++;
}
